src/legacy/core_plugins/interpreter/common/types/datatable.ts

```diff
@@ -19,75 +19,128 @@
+
+/**
+ * This type represents a `DatatableRow` in a `Datatable`.
+ */
+export type DatatableRow = Record<string, any>;
+
+/**
+ * This type represents the shape of a column in a `Datatable`.
+ */
+export interface DatatableColumn {
+  name: string;
+  type: DatatableColumnType;
+}
+
+/**
+ * A `Datatable` in Canvas is a unique structure that represents tabulated data.
+ */
+export interface Datatable {
+  type: typeof name;
+  columns: DatatableColumn[];
+  rows: DatatableRow[];
+}
+
+interface SerializedDatatable extends Datatable {
+  rows: string[][];
+}
+
+type ToRender = Render<{
+  datatable: Datatable;
+  paginate: boolean;
+  perPage: number;
+  showHeader: boolean;
+}>;
+
+export const datatable = (): ExpressionType<typeof name, Datatable, SerializedDatatable> => ({
+  name,
+  validate: dtable => {
     // TODO: Check columns types. Only string, boolean, number, date, allowed for now.
-    if (!datatable.columns) {
+    if (!dtable.columns) {
       throw new Error('datatable must have a columns array, even if it is empty');
     }
 
-    if (!datatable.rows) throw new Error('datatable must have a rows array, even if it is empty');
+    if (!dtable.rows) {
+      throw new Error('datatable must have a rows array, even if it is empty');
+    }
   },
-  serialize: datatable => {
-    const { columns, rows } = datatable;
+  serialize: dtable => {
+    const { columns, rows } = dtable;
     return {
-      ...datatable,
+      ...dtable,
       rows: rows.map(row => {
         return columns.map(column => row[column.name]);
       }),
     };
   },
-  deserialize: datatable => {
-    const { columns, rows } = datatable;
+  deserialize: dtable => {
+    const { columns, rows } = dtable;
     return {
-      ...datatable,
-      rows: rows.map(row => {
+      ...dtable,
+      rows: rows.map((row: any) => {
         return zipObject(map(columns, 'name'), row);
       }),
     };
   },
   from: {
     null: () => {
       return {
-        type: 'datatable',
+        type: name,
         rows: [],
         columns: [],
       };
     },
-    pointseries: context => {
+    pointseries: (context: PointSeries) => {
       return {
-        type: 'datatable',
+        type: name,
         rows: context.rows,
-        columns: map(context.columns, (val, name) => {
-          return { name: name, type: val.type, role: val.role };
+        columns: map(context.columns, (val, colName) => {
+          return { name: colName!, type: val.type };
         }),
       };
     },
   },
   to: {
-    render: datatable => {
+    render: (dtable): ToRender => {
       return {
         type: 'render',
         as: 'table',
         value: {
-          datatable,
+          datatable: dtable,
           paginate: true,
           perPage: 10,
           showHeader: true,
         },
       };
     },
-    pointseries: datatable => {
+    pointseries: (dtable): PointSeries => {
```

Comment:
nit: is there a reason `dtable` is not typed inline?  I'm sure it's coming from somewhere, but the fact I can't find it easily might mean an inline type that matches would be helpful.

Reply:
@clintandrewhall Yep, it's because this is already handled in the `ExpressionType` generic. We know that any function on a key in `to` is always going to cast the current type (which is passed into the generic) to something else. And we know that any function on a key in `from` is always going to return the current type (which is passed into the generic).

So TS should enforce here that the first arg to any of these `to` functions matches the `Datatable` interface

Reply:
But to your point, if that feels too "magical" it wouldn't hurt to explicitly say `dtable: Datatable`, it just isn't necessary as the generic already wires it up for you